### PR TITLE
AML-47 integrate Interval Analyser

### PR DESCRIPTION
**Pull request by tstreule**
This PR includes
- Integrate `IntervalAnalyser` into `Pipeline`
- Various minor changes at `IntervalAnalyser` class
  - Add `target` and `noise` to class attributes
- Implement basic functionality for `AmploAPI`
  - `interface.py` is the main class for user interaction
  - `platform.py` is for uploading trained models to Amplo's platform
    - **Note that the corresponding test cases are skipped since the platform is still in beta and may experience bigger changes**
  - `storage.py` is for downloading training data from Azure blob storage
  - Add environment variables (`AZURE_STORAGE_STRING ` and `AMPLO_PLATFORM_STRING`)
- Various other smaller improvements such as bug fixes and adding documentation
- Add and/or improve test cases
  - Rename `tests/Unit/AutoML` to `tests/Unit/AutoML_` in order to prevent confusion with `AutoML` folder created by pipeline

### CODE REVIEW

File: Amplo/AutoML/IntervalAnalyser.py [hunk 1]
```diff
@@ -174,32 +195,40 @@ def _parse_data(self):
...
+            dfs = pd.concat(dfs)
 
         # Extract string labels
-        self._str_labels = copy.deepcopy(dfs['labels'])
+        self._str_labels = copy.deepcopy(dfs[self.target])
 
         # Clean data
-        dp = DataProcessor(missing_values='zero', target='labels')
+        if self._pipeline is not None:
+            self._pipeline._data_processing(dfs, add_to_name='_interval')  # noqa
```
Comment (nielsuit227) [inline Amplo/AutoML/IntervalAnalyser.py:210]: Which functionality is lacking in testing?

File: Amplo/AutoML/IntervalAnalyser.py [hunk 2]
```diff
@@ -174,32 +195,40 @@ def _parse_data(self):
...
-        labels = dfs['labels']
-        dfs = dfs.drop('labels', axis=1)
+        labels = dfs[self.target]
+        dfs = dfs.drop(self.target, axis=1)
 
         # Select features -- bit hacky to avoid memory
         fp = FeatureProcessor(extract_features=False, mode='classification')
         fp.x, fp.y = dfs, labels
-        features, _ = fp._sel_gini_impurity()
+        features, _ = fp._sel_gini_impurity()  # noqa
```
Comment (nielsuit227) [inline Amplo/AutoML/IntervalAnalyser.py:231]: What functionality is missing from testing?

File: tests/Integration/test_pipeline.py [hunk 3]
```diff
@@ -169,13 +162,27 @@ def test_classification(self):
...
+        data_dir = './test_data'
+        make_interval_data(directory=data_dir)
+        # TODO: does it have to work also with n_labels=1 and n_logs=1?
+
+        # Pipeline
+        pipeline = Pipeline(grid_search_iterations=0)
+        pipeline.fit(data_dir)
+
+        # Remove dummy data
+        shutil.rmtree(data_dir)
```
Comment (nielsuit227) [inline tests/Integration/test_pipeline.py:188]: This is lacking some asserts. I guess checking if a model exists would be truly end-to-end integration test. Perhaps there's also some intermediate steps that change with the interval analyser which we can check?

File: tests/Unit/API/test_interface.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    def test_upload(self):
+        return  # TODO: remove as soon as platform is working
+
+        # Make dummy production data
+        issue_dir, kwargs = make_production_data(
+            self.sync_dir, team='Demo', machine='Charger 75kW', service='Diagnostics')
+        # Set up api and upload model
+        api = API(self.sync_dir, verbose=2)
+        api.upload_models([[issue_dir, kwargs['team'], kwargs['machine'], kwargs['service'],
+                            kwargs['issue'], kwargs['version']]])
```
Comment (nielsuit227) [inline tests/Unit/API/test_interface.py:39]: This is also lacking tests. There's quite some complex behavior in this script. I would like to see `read_latest_version`, `_iter_data` also tested.

File: tests/Unit/API/test_platform.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    def test_upload(self):
+        return  # TODO: remove as soon as platform is working
+
+        # Make dummy production data
+        issue_dir, kwargs = make_production_data(
+            self.sync_dir, team='Demo', machine='Charger 75kW', service='Diagnostics')
+        # Upload
+        sync = PlatformSynchronizer()
+        sync.upload_latest_model(issue_dir, **kwargs)
```
Comment (nielsuit227) [inline tests/Unit/API/test_platform.py:16]: This only tests that `upload_latest_model` doesn't error. 
- Good to check whether the file exists in the cloud. 
- Also needs some cleanup to make sure it doesn't stay there. 
- Also does `upload_latest_model` truly take the latest?
- What if latest locally and in the cloud are outdated? It needs to go in the right folder.

File: tests/Unit/API/test_storage.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+
+class TestStorage(APITestCase):
+
+    def test_download(self):
+        blob_dir = 'Demo/Charger 75kW/data/Diagnostics/Blocked Radiator'
+        local_dir = self.sync_dir / 'Demo/Charger 75kW/Diagnostics/data/Blocked Radiator'
+        # Set up API and synch data
+        sync = AzureSynchronizer(verbose=2)
+        sync.sync_files(blob_dir, local_dir)
```
Comment (nielsuit227) [inline tests/Unit/API/test_storage.py:12]: Also here, the functions `get_dirs`, `get_files`, and `sync_files` need there individual tests to assert good behavior. No error triggered does not mean this works well.

**Review by nielsuit227 — CHANGES_REQUESTED** [review]

File: Amplo/API/platform.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            assert get_response.status_code == 200, 'GET request has failed'
+
+            # Get all existing models
+            try:
+                # Convert string to list of dict
+                existing_models = json.loads(get_response.text)
+                if isinstance(existing_models, str):
+                    ValueError('Could not transform response.')
+            except json.decoder.JSONDecodeError as err:
+                # Empty
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/API/platform.py:95]: This means to set `existing_models` to a list of (empty) dict.
I'll adjust the comment, of course

File: Amplo/AutoML/IntervalAnalyser.py (hunk 1 above)
Comment (tstreule) [reply Amplo/AutoML/IntervalAnalyser.py:210]: Please excuse me but I'm not sure whether I understand your question

File: Amplo/AutoML/IntervalAnalyser.py (hunk 2 above)
Comment (tstreule) [reply Amplo/AutoML/IntervalAnalyser.py:231]: Please excuse me but I'm not sure whether I understand your question

File: Amplo/API/interface.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+            # Skip training model when no new data
+            if data.equals(latest_data):
+                self.print(f'Skipped training for machine {machine} as no new data was found.', 'gray')
+                continue
+
+            # Store for versioning and continue to train model
+            data.to_csv(save_dir / f'training_data_{int(time.time())}.csv')
+
+            # Train one model for each issue
+            for issue in data['labels'].unique():
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/API/interface.py]: commit caea5ce1a5fdf5856ab27a36806bdbc316ab7c1b

File: Amplo/API/platform.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+            assert isinstance(existing_models[0], dict), 'Received unexpected data type'
+
+            # Check if model exists
+            return any(model.get('name', None) == model_name for model in existing_models)
+
+        # POST when model yet doesn't exist
+        if not model_exists(data['name']):
+            post_response = requests.post(self.api_url + '/models', data=data, headers=headers)
+            # Assert that now model exists
+            assert model_exists(data['name']), f'Failed to create model - {post_response} {post_response.text}'
```
[the remark this answers is not in the document]
Comment (tstreule) [reply Amplo/API/platform.py:109]: commit caea5ce1a5fdf5856ab27a36806bdbc316ab7c1b

File: tests/Unit/API/test_interface.py (hunk 4 above)
Comment (tstreule) [reply tests/Unit/API/test_interface.py]: commit caea5ce1a5fdf5856ab27a36806bdbc316ab7c1b

File: tests/Unit/API/test_platform.py (hunk 5 above)
Comment (tstreule) [reply tests/Unit/API/test_platform.py]: commit caea5ce1a5fdf5856ab27a36806bdbc316ab7c1b

Comment (tstreule) [reply tests/Unit/API/test_platform.py]: I left this test case as is. I think it makes sense to wait until you give green light and are confident to say that the platform API is out of beta mode.

File: tests/Unit/API/test_storage.py (hunk 6 above)
Comment (tstreule) [reply tests/Unit/API/test_storage.py:28]: commit caea5ce1a5fdf5856ab27a36806bdbc316ab7c1b

**Pull request merged**